### PR TITLE
Image Customizer: dm-verity enablement.

### PR DESCRIPTION
**Pull request by liulanze**
###### Merge Checklist  

**All** boxes should be checked before merging the PR *(just tick any boxes which don't apply to this PR)*
- [ ] The toolchain has been rebuilt successfully (or no changes were made to it)
- [ ] The toolchain/worker package manifests are up-to-date
- [ ] Any updated packages successfully build (or no packages were changed)
- [ ] Packages depending on static components modified in this PR (Golang, `*-static` subpackages, etc.) have had their `Release` tag incremented.
- [ ] Package tests (%check section) have been verified with RUN_CHECK=y for existing SPEC files, or added to new SPEC files
- [ ] All package sources are available
- [ ] cgmanifest files are up-to-date and sorted (`./cgmanifest.json`, `./toolkit/scripts/toolchain/cgmanifest.json`, `.github/workflows/cgmanifest.json`)
- [ ] LICENSE-MAP files are up-to-date (`./SPECS/LICENSES-AND-NOTICES/data/licenses.json`, `./SPECS/LICENSES-AND-NOTICES/LICENSES-MAP.md`, `./SPECS/LICENSES-AND-NOTICES/LICENSE-EXCEPTIONS.PHOTON`)
- [ ] All source files have up-to-date hashes in the `*.signatures.json` files
- [ ] `sudo make go-tidy-all` and `sudo make go-test-coverage` pass
- [ ] Documentation has been updated to match any changes to the build system
- [ ] Ready to merge

---

###### Summary 

What does the PR accomplish, why was it needed?
- This PR initialized the logic enabling dm-verity in MIC for the images want root / user partition protection.
- Please note this PR is not finished yet, for instance unit tests are working in progress, but the code style / syntax can be reviewed in the meantime.

###### Change Log  

- Change
- Change
- Change

###### Does this affect the toolchain?  

**NO**

###### Associated issues  

- #xxxx

###### Links to CVEs  
- https://nvd.nist.gov/vuln/detail/CVE-YYYY-XXXX

###### Test Methodology

- Local

### CODE REVIEW

File: toolkit/tools/pkg/imagecustomizerlib/customizeutils.go [hunk 1]
```diff
@@ -84,6 +84,11 @@ func doCustomizations(buildDir string, baseConfigPath string, config *imagecusto
 		return err
 	}
 
+	err = enableVerityPartition(config.SystemConfig.Verity, imageChroot)
```
Comment (cwize1) [inline toolkit/tools/pkg/imagecustomizerlib/customizeutils.go:87]: I assume that dm-verity (like encryption) requires the boot partition to be separate from the rootfs partition. If this is the case, then a check should be added to make sure that is actually the case.

File: toolkit/tools/pkg/imagecustomizerlib/customizeverity.go [hunk 2]
```diff
@@ -0,0 +1,182 @@
...
+	const grubCfgPath = "/boot/grub2/grub.cfg"
+	const cmdlineTemplate = "rd.systemd.verity=1 roothash=%s systemd.verity_root_data=%s systemd.verity_root_hash=%s systemd.verity_root_options=restart-on-corruption rootflags=noload"
+	const rootDevicePattern = "set rootdevice=PARTUUID="
+    const replacement = "set rootdevice=/dev/mapper/root"
+
+	newArgs := fmt.Sprintf(cmdlineTemplate, rootHash, Verity.VerityDevice, Verity.HashDevice)
+
+	var updatedLines []string
+
+	err := imageChroot.Run(func() error {
```
Comment (cwize1) [inline toolkit/tools/pkg/imagecustomizerlib/customizeverity.go:151]: If you are just reading and writing files, you don't need to do this under chroot.

File: toolkit/tools/pkg/imagecustomizerlib/customizeverity.go [hunk 3]
```diff
@@ -0,0 +1,182 @@
...
+		return ioutil.WriteFile(cfgPath, []byte(output), 0644)
+	})
+
+	return nil
+}
+
+func calculateRootHash(Verity imagecustomizerapi.Verity, imageChroot *safechroot.Chroot) (string, error) {
+	var err error
+	var rootHash string
+	cmd := fmt.Sprintf("veritysetup -v --debug format %s %s | grep 'Root hash: ' | awk '{print $3}'", Verity.VerityDevice, Verity.HashDevice)
```
Comment (cwize1) [inline toolkit/tools/pkg/imagecustomizerlib/customizeverity.go:125]: Don't use external tools to process output. Process the output directly in Go.

File: toolkit/tools/pkg/imagecustomizerlib/customizeverity.go [hunk 4]
```diff
@@ -0,0 +1,182 @@
...
+	}
+
+	return nil
+}
+
+func buildDracutModule(dracutModuleName string, imageChroot *safechroot.Chroot) error {
+	var err error
+
+	buildDracutModuleArgs := []string{
+		"dracut", "-f", "-a",
```
Comment (cwize1) [inline toolkit/tools/pkg/imagecustomizerlib/customizeverity.go:57]: Calling `dracut` directly can cause a lot of problems. The correct way to handle this is to add the module by creating a new config file under `/etc/dracut.conf.d/`. After that, reinstalling the `dracut` package should trigger the update to the initramfs file.

**Comment by liulanze** [comment]
Close this draft PR and migrate to the published one: https://github.com/microsoft/CBL-Mariner/pull/6809/files. However, will address comments received in this PR, thanks.

**Pull request closed without merging**